ksqldb-engine/src/main/java/io/confluent/ksql/engine/QueryMonitor.java

```diff
@@ -0,0 +1,215 @@
+  private static final long BASE_WAITING_TIME_MS = 10000;
+  private static final int SHUTDOWN_TIMEOUT_MS = 5000;
+
+  private final Ticker ticker;
+  private final long retryBackoffMaxMs;
+  private final KsqlEngine ksqlEngine;
+  private final ExecutorService executor;
+  private final Map<QueryId, RetryEvent> queriesRetries = new HashMap<>();
+
+  private volatile boolean closed = false;
+
+  public QueryMonitor(final KsqlConfig ksqlConfig, final KsqlEngine ksqlEngine) {
+    this(
+        ksqlConfig,
+        ksqlEngine,
+        Executors.newSingleThreadExecutor(r -> new Thread(r, "QueryMonitor")),
+        CURRENT_TIME_MILLIS_TICKER
+    );
+  }
+
+  @VisibleForTesting
+  QueryMonitor(
+      final KsqlConfig ksqlConfig,
+      final KsqlEngine ksqlEngine,
+      final ExecutorService executor,
+      final Ticker ticker
+  ) {
+    this.retryBackoffMaxMs = ksqlConfig.getLong(KsqlConfig.KSQL_QUERY_RETRY_BACKOFF_MAX_MS);
+    this.ksqlEngine = ksqlEngine;
+    this.executor = executor;
+    this.ticker = ticker;
+  }
+
+  public void start() {
+    executor.execute(new Runner());
+    executor.shutdown();
+  }
+
+  @Override
+  public void close() {
+    closed = true;
+
+    try {
+      executor.awaitTermination(SHUTDOWN_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    } catch (final InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
+  }
+
+  boolean isClosed() {
+    return closed;
+  }
+
+  void restartFailedQueries() {
+    ksqlEngine.getPersistentQueries().stream()
+        .forEach(query -> {
+          final QueryId queryId = query.getQueryId();
+          final KafkaStreams.State queryState = query.getState();
+
+          // If the query was restarted previously, check if it needs another restart; or if the
+          // query is now up and running, then remove it from the restart list.
+          if (queriesRetries.containsKey(queryId)) {
+            final RetryEvent retryEvent = queriesRetries.get(queryId);
+            if (ticker.read() > retryEvent.nextRestartTimeMs()) {
+              if (queryState == KafkaStreams.State.ERROR) {
+                // Retry again if it's still in ERROR state
+                retryEvent.restart();
+              } else {
+                // Query is not in ERROR state anymore.
+                queriesRetries.remove(queryId);
+              }
+            }
+          } else if (queryState == KafkaStreams.State.ERROR) {
+            // Restart new query in ERROR state, and add it to the list of retries.
+            final RetryEvent retryEvent = new RetryEvent(
+                ksqlEngine, queryId, BASE_WAITING_TIME_MS, retryBackoffMaxMs, ticker);
+
+            queriesRetries.put(queryId, retryEvent);
+            retryEvent.restart();
+          }
+        });
+  }
+
+  private class Runner implements Runnable {
+    @Override
+    public void run() {
+      LOG.info("KSQL query monitor started.");
+
+      while (!closed) {
+        try {
+          restartFailedQueries();
```

Comment:
instead of having the try/catch around the entire batch should we try/catch each individual query restart? that way if there's some issue causing the very first one to restart we won't just keep trying only the very first one

Reply:
Done. I added the try/catch in the `RetryEvent`.